docker_cli/commands.py

```diff
@@ -66,6 +67,61 @@ def get_archive(
     return Image.build(tag=img_tag, dockerfile_string=dockerfile, no_cache=True)
 
 
+def insert(tag: str, base: str, directory: str | os.PathLike[str]):
+    """
+    Builds a Docker image with the contents of the given directory copied onto it.
+
+    The directory path on the image has the same name as the topmost directory
+    of the given path. e.g. giving path "/tmp/dir/subdir" will result in the contents of
+    this path being saved in "/subdir" on the generated image.
+
+    This Dockerfile also places the working directory of the image inside of the copied
+    directory.
+
+    Parameters
+    ----------
+    tag : str
+        The image tag.
+    base : str
+        The base image tag.
+    directory : path-like
+        The directory to be copied.
+
+    Returns
+    -------
+    Image
+        The generated image.
+    """
+
+    prefix = universal_tag_prefix()
+    img_tag = tag if tag.startswith(prefix) else f"{prefix}-{tag}"
+
+    dir_str = str(directory)
+
+    # The absolute path of the given directory will be the build context.
+    # This is necessary because otherwise docker may be unable to find the directory if
+    # the build context is at the current working directory.
+    path_absolute = os.path.abspath(str(dir_str))
+    # Additionally, the top directory of the given path will be the name of the
+    # directory in the image.
+    if os.path.isdir(dir_str):
+        target_dir = os.path.basename(path_absolute)
+    else:
+        target_dir = os.path.basename(os.path.dirname(path_absolute))
```

Comment:
Let's change the `else` branch here to raise an exception instead.

The input `directory` argument of this function is explicitly intended to be an existing directory on the host. If the user passed in something that wasn't a valid directory, we should raise an exception here to let them know about their mistake.